### ISSUE TITLE
ConfigureAwait(this) shouldn't run the continuation if the owner is destroyed.

### ISSUE THREAD

**Issue by mikehofm** [muckSponge/UnityAsync#1]
Love your library!

Say you have some code like this:
```
async void Start()
{
    await Await.Seconds(5).ConfigureAwait(this);
    transform.position = new Vector3(1, 2, 3);  // will crash if object was destroyed
}
```

Currently, if the GameObject or Component is destroyed, the continuation runs immediately, which would cause a crash. I think it would be more intuitive if the continuation were cancelled instead. This is probably what the user wants most of the time and is consistent with how coroutines work.

**Comment by muckSponge** [comment]
Yeah, I think you're right so I've flipped the logic for continuation cancellations. Thanks for testing, let me know if it fixes the issue :)

**Comment by mikehofm** [comment]
Yep, that fixes it. For cancellation tokens, I'm not sure what the right answer is. I can see it either way, depending on the situation.

If you have something like this, you may want it to continue:
```
async void ShowLeaderboards()
{
    await RefreshLeaderboardsFromInternet().ConfigureAwait(cancellationToken);
    DisplayLeaderboards();  // leaderboards may be slightly out of date if the refresh was cancelled
}
```
While in other cases, you'd want it to cancel:
```
var importantData = await GetImportantData.ConfigureAwait(cancellationToken);
ProcessImportantData(importantData);
```
The usual pattern with Tasks would be to call cancellationToken.ThrowIfCancellationRequested(). Then the client code can wrap the await in a try-catch and decide what to do if it's cancelled.

**Comment by muckSponge** [comment]
That makes sense. I'm struggling to work out how to implement this kind of behaviour for `Continuation<IAwaitInstruction>`.

EDIT: Removed my previous ramblings because I think I've solved the problem. `Continuation<IAwaitInstruction>` is a struct, which limits how we can work with exceptions. At various points the struct will be copied and unfortunately, that means when the continuation delegate is run, and `GetResult` is invoked, it has an outdated copy of the struct, so we can't store an `Exception` field in the struct and rethrow in `GetResult`, because there will never be an exception there. My solution is to just use a static `Exception` field and clear it in `GetResult` before rethrowing it. I think this works because all `Continuation<IAwaitInstruction>`s are processed on the main thread and the continuation is run straight after `GetResult` is invoked, so there's no race condition on the static exception field, because no other continuations have been evaluated yet.

This gives us our `OperationCancelledException` so we can do something like:

```c#
async void Start()
{
	float t = Time.time;
	
	// cancel after 500ms
	var source = new CancellationTokenSource(500);

	try
	{
		// wait for 2000ms unless cancelled
		Await.Seconds(2f).ConfigureAwait(source.Token);
	}
	catch(Exception e)
	{
		Debug.LogException(e);
	}
	
	Debug.Log(Time.time - t);
}
```

That will print an `OperationCancelledException` to console after ~500ms.

Now the question is, should we throw an exception if the owner of the await instruction is destroyed, or perhaps add that as a `ConfigureAwait` option? The more options we add, the more fields we need to store in the `Continuation<IAwaitInstruction>` struct and the more expensive it becomes due to struct copies.

**Comment by mikehofm** [comment]
So my instinct is that once an object is destroyed, you rarely want to continue with the continuation. Your code is probably going to use some member of the owning object and crash; if the code had no dependencies on that object, it probably wouldn't be in the object to begin with (or at least, you wouldn't have called ConfigureAwait(this)). Of course, it's hard to predict what use cases people might have.

I think the static Exception variable is safe here, but I don't know enough about async/await to be confident.

One unrelated thing I noticed: in ContinuationProcessor, you don't clean up the completed IContinuations. This could prevent any references held by the IContinuation from being garbage collected. You could probably add `Array.Clear(currentQueue, 0, count);` at the end of Process().

**Comment by muckSponge** [comment]
Yes, I'm unable to come up with a logical situation where someone would tie it to an owner that doesn't await on it or at least have some dependencies with the awaiting object so probably no point complicating things with an exception or exception option.

I've resolved the potential managed memory leak in 5430f351c9a06377d61018a946dcbb24c55418b4.
